### PR TITLE
[PM-5735] Create kdf Service

### PR DESCRIPTION
**Pull request by ike-kottlowski**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

[PM-5735](https://bitwarden.atlassian.net/browse/PM-5735)

## Objective

Create a service to fetch new `KdfConfig` object. At the same time remove Kdf References from the `StateService`

## Code changes

Primary code changes are renaming and object accessor adjustments due to combining the `KdfType` into the `KdfConfig` object. 

There are some DI edits to ensure proper injection across clients.

- **kdf-config-service.factory.ts**: Added factory to support DI for Browser.
- **browser/src/background/main.background.ts**: Updated DI for various services depending on `KdfConfig` and removing dependency from `StateService` in some cases.
- **crypto.service.ts**: removed reference to `kdfType` in quite a few methods. The type is now contained in the KdfConfig.
  - Also moved the `validateKdfConfig` to the new `KdfService`
- **kdfConfig.service.ts**: created abstraction and implementation classes for get/set KdfConfig from `StateProvider`. Moved `validateKdfConfig` from `CryptoService` to here.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

[PM-5735]: https://bitwarden.atlassian.net/browse/PM-5735?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/cli/src/bw.ts [hunk 1]
```diff
@@ -232,6 +233,7 @@ export class Main {
   stateEventRunnerService: StateEventRunnerService;
   biometricStateService: BiometricStateService;
   billingAccountProfileStateService: BillingAccountProfileStateService;
+  kdfConfigService: KdfConfigService;
```
[the remark this answers is not in the document]
Comment (ike-kottlowski) [reply apps/cli/src/bw.ts]: Addressed and tested the CLI login successfully.

File: libs/common/src/auth/models/domain/kdf-config.ts [hunk 2]
```diff
@@ -1,11 +1,69 @@
...
+  ARGON2_ITERATIONS,
+  ARGON2_MEMORY,
+  ARGON2_PARALLELISM,
+  KdfType,
+  PBKDF2_ITERATIONS,
+} from "../../../platform/enums/kdf-type.enum";
+
+export type KdfConfig = PBKDF2KdfConfig | Argon2KdfConfig;
+
+export class PBKDF2KdfConfig {
```
[the remark this answers is not in the document]
Comment (ike-kottlowski) [reply libs/common/src/auth/models/domain/kdf-config.ts:19]: Added some comments.

**Review by jlf0dev — APPROVED** [review]

**Review by shane-melton — APPROVED** [review]
Vault change looks good!

**Review by r-tome — APPROVED** [review]

**Review by djsmith85 — APPROVED** [review]
Changes are looking great and approving for tools!

**Review by MGibson1 — CHANGES_REQUESTED** [review]

File: libs/common/src/state-migrations/migrations/59-move-kdf-config-to-state-provider.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      await helper.set(userId, account);
+    }
+    await Promise.all([...accounts.map(({ userId, account }) => migrateAccount(userId, account))]);
+  }
+
+  async rollback(helper: MigrationHelper): Promise<void> {
+    const accounts = await helper.getAccounts<ExpectedAccountType>();
+    async function rollbackAccount(userId: string, account: ExpectedAccountType): Promise<void> {
+      const kdfConfig: KdfConfig = await helper.getFromUser(userId, kdfConfigKeyDefinition);
+
```
[the remark this answers is not in the document]
Comment (ike-kottlowski) [reply libs/common/src/state-migrations/migrations/59-move-kdf-config-to-state-provider.ts:65]: How would I make this more defensive?

[the remark this answers is not in the document]
Comment (MGibson1) [reply libs/common/src/state-migrations/migrations/59-move-kdf-config-to-state-provider.ts:65]: By guarding against null `account` or null `account.profile`

```typescript
      if (account && kdfConfig) {
        account.keys = Object.assign(account.profile ?? {}, {
          kdfIterations = kdfConfig.iterations;
          kdfType = kdfConfig.kdfType;
          kdfMemory = kdfConfig.memory;
          kdfParallelism = kdfConfig.parallelism;
        });
        await helper.set(userId, account);
      }
```

**Review by MGibson1 — APPROVED** [review]

**Review by djsmith85 — APPROVED** [review]

**Review by r-tome — APPROVED** [review]

**Review by jlf0dev — APPROVED** [review]

**Review by shane-melton — APPROVED** [review]

**Pull request merged**